### PR TITLE
refactor: user experience improvement after fine-tuning

### LINKED ISSUE #511: (Refactoring) Integrate commons packages to the Client

### PR DESCRIPTION
**Pull request by bwanglzu**
Closes #511 #514 

1. take out callback stubs, dependent on `finetuner-stubs`.  DONE
2. take out model stubs, dependent on `finetuner-stubs`.  DONE
3. add `finetuner-commons` as a soft dependency: `pip install "finetuner[full]"` DONE
4. provide extra functionality based on `finetuner-commons`, includs `get_model` and `preprocess_and_collate` DONE:

```python
# after fine-tuning
artifact = run.save_artifact('my-model')
# artifacts include model and preprocess/collate function
# re-build the model 
model = finetuner.get_model(artifact=artifact)
# preprocess, collate and encode
finetuner.encode(model=model, data=da)
```

---

- [x] This PR references an open issue
- [x] I have added a line about this change to CHANGELOG

### CODE REVIEW

**Comment by bwanglzu** [comment]
related PR in core: https://github.com/jina-ai/finetuner-core/pull/307

**Comment by guenthermi** [comment]
For the clip models the stubs used here before are quite different from the stubs in finetuner-core (and finetuner-commons), i.e., there was only one stub instead of two for clip-text and clip-vision. This should be taken care of, when the code is refactored, e.g., in the get_row method:
https://github.com/jina-ai/finetuner/blob/5b904f2ecd231ddf2cd80ba8d45985dcb06b11f0/finetuner/model/__init__.py#L9-L17

**Comment by bwanglzu** [comment]
waiting for PR https://github.com/jina-ai/finetuner-core/pull/308

File: setup.cfg [hunk 1]
```diff
@@ -6,6 +6,6 @@ version = 0.5.2
 # E203 is whitespace before ':' - which occurs in numpy slicing, e.g. in
 #     dists[2 * i : 2 * i + 2, :]
 # W503 is line break before binary operator - happens when black splits up lines
-ignore = E203, W503
+ignore = E203, W503, F405, F403
```
Comment (gmastrapas) [inline setup.cfg:9]: what do these ignore?

File: setup.py [hunk 2]
```diff
@@ -69,9 +61,26 @@
         long_description_content_type='text/markdown',
         zip_safe=False,
         setup_requires=_setup_requires,
-        install_requires=_main_deps,
+        install_requires=[
+            'docarray[common]>=0.13.31',
+            'finetuner-stubs==0.0.1b1',
```
Comment (gmastrapas) [inline setup.py]: we dont have an official version yet?

File: finetuner/callback/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from stubs.callback import *  # noqa F401
```
Comment (gmastrapas) [inline finetuner/callback/__init__.py:1]: why do you create a directory here? why not a `callback.py` module?

File: finetuner/model/__init__.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from stubs.model import *  # noqa F401
```
Comment (gmastrapas) [inline finetuner/model/__init__.py:1]: same here, why not a `model.py` module?

File: finetuner/run.py [hunk 5]
```diff
@@ -74,12 +75,22 @@ def logs(self) -> str:
             experiment_name=self._experiment_name, run_name=self._name
         )
 
-    def stream_logs(self) -> Iterator[str]:
+    def stream_logs(self, interval: int = 5) -> Iterator[str]:
```
Comment (gmastrapas) [inline finetuner/run.py:78]: let's document the interval argument here

File: finetuner/__init__.py [hunk 6]
```diff
@@ -39,16 +40,26 @@ def list_callbacks() -> Dict[str, callback.CallbackStubType]:
...
-        if (
-            inspect.isclass(obj)
-            and issubclass(obj, models._ModelStub)
-            and obj != models._ModelStub
-        )
-    }
+def _list_models() -> Dict[str, model.ModelStubType]:
+    from stubs import model as model_stub
+
+    rv = {}
```
Comment (gmastrapas) [inline finetuner/__init__.py:46]: what is rv?

File: finetuner/__init__.py [hunk 7]
```diff
@@ -39,16 +40,26 @@ def list_callbacks() -> Dict[str, callback.CallbackStubType]:
...
+    for name, stub in members:
+        if name != 'MLPStub' and not name.startswith('_') and type(stub) != type:
+            rv[name] = stub
+    return rv
+
+
+def _build_name_stub_map() -> Dict[str, model.ModelStubType]:
+    from stubs import model as model_stub
+
+    rv = {}
```
Comment (gmastrapas) [inline finetuner/__init__.py:57]: what is rv?

File: finetuner/__init__.py [hunk 8]
```diff
@@ -161,7 +172,7 @@ def fit(
     :param learning_rate: learning rate for the optimizer.
     :param epochs: Number of epochs for fine-tuning.
     :param batch_size: Number of items to include in a batch.
-    :param callbacks: List of callback stub objects. See the `finetuner.callback`
+    :param callbacks: List of callback stub objects.`
```
Comment (gmastrapas) [inline finetuner/__init__.py:175]: ` char in the end

File: finetuner/__init__.py [hunk 9]
```diff
@@ -296,3 +307,82 @@ def get_token() -> str:
     :return: user token as string object.
     """
     return ft.get_token()
+
+
+def get_model(
```
Comment (gmastrapas) [inline finetuner/__init__.py:312]: I dont like the get_model name, maybe `build_engine`?

File: finetuner/__init__.py [hunk 10]
```diff
@@ -296,3 +307,82 @@ def get_token() -> str:
...
+    :param token: A Jina authentication token required for pulling artifacts from
+        Hubble. If not provided, the Hubble client will try to find one either in a
+        local cache folder or in the environment.
+    :param batch_size: Incoming documents are fed to the graph in batches, both to
+        speed-up inference and avoid memory errors. This argument controls the
+        number of documents that will be put in each batch.
+    :param select_model: Finetuner run artifacts might contain multiple models. In
+        such cases you can select which model to deploy using this argument. For CLIP
+        fine-tuning, you can choose either `clip-vision` or `clip-text`.
+    :param gpu: if specified to True, use cuda device for inference.
```
Comment (gmastrapas) [inline finetuner/__init__.py:335]: logging level is missing from the docstring

File: setup.py (hunk 2 above)
Comment (bwanglzu) [reply setup.py]: until we release and trigger core CD

File: finetuner/__init__.py (hunk 9 above)
Comment (bwanglzu) [reply finetuner/__init__.py:312]: no, think about when we support pytorch, it's not engine anymore

File: finetuner/__init__.py (hunk 6 above)
Comment (bwanglzu) [reply finetuner/__init__.py:46]: this is `return_value`

**Review by gmastrapas — APPROVED** [review]

File: finetuner/__init__.py [hunk 11]
```diff
@@ -296,3 +304,85 @@ def get_token() -> str:
...
+    token: Optional[str] = None,
+    batch_size: int = 32,
+    select_model: Optional[str] = None,
+    gpu: bool = False,
+    logging_level: str = 'WARNING',
+):
+    """Re-build the model based on the model inference session with ONNX.
+
+    :param artifact: Specify a finetuner run artifact. Can be a path to a local
+        directory, a path to a local zip file or a Hubble artifact ID. Individual
```
Comment (guenthermi) [inline finetuner/__init__.py:320]:
```suggestion
        directory, a path to a local zip file, or a Hubble artifact ID. Individual
```

File: finetuner/__init__.py [hunk 12]
```diff
@@ -296,3 +304,85 @@ def get_token() -> str:
...
+    )
+    return ort_engine
+
+
+def encode(
+    model,
+    data: DocumentArray,
+    batch_size: int = 32,
+):
+    """Process, collate and encode the `DocumentArray` with embeddings.
```
Comment (guenthermi) [inline finetuner/__init__.py:366]: I think it is "[Pre-]process"

File: finetuner/__init__.py [hunk 13]
```diff
@@ -296,3 +304,85 @@ def get_token() -> str:
...
+    :param data: The `DocumentArray` object to be encoded.
+    :param batch_size: Incoming documents are fed to the graph in batches, both to
+        speed-up inference and avoid memory errors. This argument controls the
+        number of documents that will be put in each batch.
+    :returns: `DocumentArray` filled with embeddings.
+
+    ..Note::
+      please install finetuner[full] to include all the dependencies.
+    """
+    for i, batch in enumerate(data.batch(batch_size)):
```
Comment (guenthermi) [inline finetuner/__init__.py:380]: where do you need the `i`

**Review by guenthermi — COMMENTED** [review]
Added some minor comments

**Pull request merged**